braindecode/models/modules.py

```diff
@@ -528,102 +529,289 @@ def extra_repr(self) -> str:
-        Returns:
-            Tensor: Output signal of dimension (..., time).
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
         """
-        if self.orig_freq == self.new_freq:
-            return waveform
-        return _apply_sinc_resample_kernel(
-            waveform, self.orig_freq, self.new_freq, self.gcd, self.kernel, self.width
-        )
+        Apply the filter bank to the input signal.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input tensor of shape (batch_size, time_points).
+
+        Returns
+        -------
+        torch.Tensor
+            Filtered output tensor of shape (batch_size, n_bands, filtered_time_points).
+        """
+        # Initialize a list to collect filtered outputs
+        n_bands = self.n_bands
+        output = []
+
+        for band_idx in range(n_bands):
+            if not self.method_iir:
+                filtered = self._apply_fir(
+                    x, self.filts[f"{band_idx}_band"], self.n_chans
+                )
+            else:
+                filtered = self._apply_irr(
+                    x, self.filts[f"{band_idx}_band"], self.n_bands
+                )
+
+            output.append(filtered)
+
+        # Shape: (batch_size, n_bands, nchans, time_points)
+        output = torch.cat(output, dim=1)
+        return output
+
+    @staticmethod
+    def _create_band_filters(
+        start, end, num_bands
+    ) -> list[Tuple[float | int, float | int]]:
+        total_numbers = end - start + 1
+        base_interval_size = total_numbers // num_bands
+
+        intervals = []
+        current_start = start
+
+        for i in range(num_bands):
+            # For the last interval, add any extra values
+            if i == num_bands - 1:
+                current_end = end
+            else:
+                current_end = current_start + base_interval_size - 1
+            intervals.append((current_start, current_end))
+            current_start = current_end + 1
+
+        # Convert intervals to a list of tuples (low_freq, high_freq)
+        band_filters = [(low, high) for low, high in intervals]
+        return band_filters
+
+    @staticmethod
+    def _apply_fir(x, filt_b, n_chans) -> Tensor:
+        # Shape: (nchans, filter_length)
+
+        # Expand to (nchans, filter_length)
+        filt_expanded = filt_b.unsqueeze(0).repeat(n_chans, 1).unsqueeze(0)
+
+        # I think it will only work with FIR, check with MNE experts.
+        filtered = fftconvolve(
+            x, filt_expanded, mode="same"
+        )  # Shape: (batch_size, nchans, time_points)
+
+        # Add band dimension
+        # Shape: (batch_size, 1, nchans, time_points)
+        filtered = filtered.unsqueeze(1)
+        # returning the filtered
+        return filtered
+
+    @staticmethod
+    def _apply_irr(x: Tensor, filter, n_bands: int) -> Tensor:
+        x = x.unsqueeze(2).repeat(1, 1, n_bands, 1)
+        x = filtfilt(x, filter["a"], filter["a"], clamp=False)
```

Comment:
All this should be tested to check that the filtering is done right.
```suggestion
        x = filtfilt(x, filter["a"], filter["b"], clamp=False)
```